android/src/main/java/com/oblador/keychain/KeychainModule.java

```diff
@@ -13,291 +20,586 @@
+    try {
+      boolean fingerprintAuthAvailable = isFingerprintAuthAvailable();
+
+      if (fingerprintAuthAvailable) {
+        promise.resolve(FINGERPRINT_SUPPORTED_NAME);
+      } else {
+        promise.resolve(null);
+      }
+    } catch (Exception e) {
+      Log.e(KEYCHAIN_MODULE, e.getMessage(), e);
+
+      promise.reject(Errors.E_SUPPORTED_BIOMETRY_ERROR, e);
+    } catch (Throwable fail) {
+      Log.e(KEYCHAIN_MODULE, fail.getMessage(), fail);
+
+      promise.reject(Errors.E_UNKNOWN_ERROR, fail);
+    }
+  }
+  //endregion
+
+  //region Implementation
+
+  /** Is provided access control string matching biometry use request? */
+  public static boolean getUseBiometry(@Nullable final String accessControl) {
+    return accessControl != null
+      && (accessControl.equals(ACCESS_CONTROL_BIOMETRY_ANY)
+      || accessControl.equals(ACCESS_CONTROL_BIOMETRY_CURRENT_SET));
+  }
+
+  private void addCipherStorageToMap(@NonNull final CipherStorage cipherStorage) {
+    cipherStorageMap.put(cipherStorage.getCipherStorageName(), cipherStorage);
+  }
+
+  /**
+   * Extract credentials from current storage. In case if current storage is not matching
+   * results set then executed migration.
+   */
+  @NonNull
+  private DecryptionResult decryptCredentials(@NonNull final String alias,
+                                              @NonNull final CipherStorage current,
+                                              @NonNull final ResultSet resultSet)
+    throws CryptoFailedException, KeyStoreAccessException {
+    final String storageName = resultSet.cipherStorageName;
+
+    // The encrypted data is encrypted using the current CipherStorage, so we just decrypt and return
+    if (storageName.equals(current.getCipherStorageName())) {
+      final DecryptionResultHandler handler = getInteractiveHandler(current);
+      current.decrypt(handler, alias, resultSet.username, resultSet.password, ANY);
+
+      CryptoFailedException.reThrowOnError(handler.getError());
+
+      if (null == handler.getResult()) {
+        throw new CryptoFailedException("No decryption results and no error. Something deeply wrong!");
+      }
+
+      return handler.getResult();
+    }
 
-    @Override
-    public String getName() {
-        return KEYCHAIN_MODULE;
+    // The encrypted data is encrypted using an older CipherStorage, so we need to decrypt the data first, then encrypt it using the current CipherStorage, then store it again and return
+    final CipherStorage oldStorage = getCipherStorageByName(storageName);
+    if (null == oldStorage) {
+      throw new KeyStoreAccessException("Wrong cipher storage name: " + storageName);
     }
 
-    public KeychainModule(ReactApplicationContext reactContext) {
-        super(reactContext);
-        prefsStorage = new PrefsStorage(reactContext);
+    // decrypt using the older cipher storage
+    final DecryptionResult decryptionResult = oldStorage.decrypt(
+      alias, resultSet.username, resultSet.password, ANY);
```

Comment:
To follow code style used in this library:
```suggestion
      alias, resultSet.username, resultSet.password, SecurityLevel.ANY);
```